File: 3160-find-the-number-of-distinct-colors-among-the-balls/3160-find-the-number-of-distinct-colors-among-the-balls.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        unordered_map<int, int> colorCount;
        unordered_map<int, int> prevColor;
        int n = queries.size();
        vector<int> res(n, 0);
        for (int i = 0; i < n; i++) {
            int ball = queries[i][0];
            int color = queries[i][1];
            if (prevColor.count(ball)) {
                int prev = prevColor[ball];
                colorCount[prev]--;
                if (colorCount[prev] == 0) colorCount.erase(prev);   
            }
            prevColor[ball] = color;
            colorCount[color]++;
            res[i] = colorCount.size();
        }
        return res;
    }
};






// MLE
// class Solution {
// public:
//     vector<int> queryResults(int limit, vector<vector<int>>& queries) {
//         int n = queries.size();
//         unordered_map<int, int> mp;
//         vector<int> arr(limit+1, -1);
//         vector<int> res(n);
//         for (int i = 0; i < n; i++) {
//             int ball = queries[i][0];
//             int color = queries[i][1];
//             if (arr[ball] != -1) {
//                 int prevColor = arr[ball];
//                 mp[prevColor]--;
//                 if (mp[prevColor] == 0) {
//                     mp.erase(prevColor);
//                 }
//             }
//             arr[ball] = color;
//             mp[color]++;
//             res[i] = mp.size();
//         }
//         return res;
//     }
// };

File: LeetCode/3160-find-the-number-of-distinct-colors-among-the-balls/3160-find-the-number-of-distinct-colors-among-the-balls.java
class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        Map<Integer, Integer> colorCount = new HashMap<>();
        Map<Integer, Integer> prevColor = new HashMap<>();
        int n = queries.length;
        int[] res = new int[n];

        for (int i = 0; i < n; i++) {
            int ball = queries[i][0];
            int color = queries[i][1];

            if (prevColor.containsKey(ball)) {
                int prev = prevColor.get(ball);
                colorCount.put(prev, colorCount.get(prev) - 1);
                if (colorCount.get(prev) == 0) {
                    colorCount.remove(prev);
                }
            }

            prevColor.put(ball, color);
            colorCount.put(color, colorCount.getOrDefault(color, 0) + 1);
            res[i] = colorCount.size();
        }

        return res;
    }
}